code/datums/EPv2.dm

```diff
@@ -0,0 +1,142 @@
+		var/addr_1 = hexadecimal_to_EPv2(raw_address)
+		new_address = "[addr_0]:[addr_1]"
+		string = "[string]0" //If we did get a collision, this should make the next attempt not have one.
+	address = new_address
+	GLOB.all_exonet_connections |= src
+
+
+// Proc: make_arbitrary_address()
+// Parameters: 1 (new_address - the desired address)
+// Description: Allocates that specific address, if it is available.
+/datum/exonet_protocol/proc/make_arbitrary_address(var/new_address)
+	if(new_address)
+		if(new_address == find_address(new_address) )	//Collision test.
+			return FALSE
+		address = new_address
+		GLOB.all_exonet_connections += src
+		return TRUE
+
+// Proc: hexadecimal_to_EPv2()
+// Parameters: 1 (hex - a string of hexadecimals to convert)
+// Description: Helper proc to add colons to a string in the right places.
+/proc/hexadecimal_to_EPv2(var/hex)
+	if(!hex)
+		return null
+	var/addr_1 = copytext(hex,1,5)
+	var/addr_2 = copytext(hex,5,9)
+	var/addr_3 = copytext(hex,9,13)
+	var/addr_4 = copytext(hex,13,17)
+	var/new_address = "[addr_1]:[addr_2]:[addr_3]:[addr_4]"
+	return new_address
+
+
+// Proc: remove_address()
+// Parameters: None
+// Description: Deallocates the address, freeing it for use.
+/datum/exonet_protocol/proc/remove_address()
+	address = ""
+	GLOB.all_exonet_connections -= src
+
+
+// Proc: find_address()
+// Parameters: 1 (target_address - the desired address to find)
+// Description: Searches the global list GLOB.all_exonet_connections for a specific address, and returns it if found, otherwise returns null.
+/datum/exonet_protocol/proc/find_address(var/target_address)
+	for(var/datum/exonet_protocol/exonet in GLOB.all_exonet_connections)
+		if(exonet.address == target_address)
+			return exonet.address
+	return null
+
+// Proc: get_atom_from_address()
+// Parameters: 1 (target_address - the desired address to find)
+// Description: Searches an address for the atom it is attached for, otherwise returns null.
+/datum/exonet_protocol/proc/get_atom_from_address(var/target_address)
+	for(var/datum/exonet_protocol/exonet in GLOB.all_exonet_connections)
+		if(exonet.address == target_address)
+			return exonet.holder
+	return null
+
+// Proc: send_message()
+// Parameters: 3 (target_address - the desired address to send the message to, data_type - text stating what the content is meant to be used for,
+// 		content - the actual 'message' being sent to the address)
+// Description: Sends the message to target_address, by calling receive_message() on the desired datum.  Returns true if the message is recieved.
+/datum/exonet_protocol/proc/send_message(var/target_address, var/data_type, var/content)
+	if(!address)
+		return FALSE
+	var/obj/machinery/exonet_node/node = get_exonet_node()
+	if(!node) // Telecomms went boom, ion storm, etc.
+		return FALSE
+	if(!node.on||node.stat)
+		return FALSE
+	for(var/datum/exonet_protocol/exonet in GLOB.all_exonet_connections)
```

Comment:
this should be a typeless loop

Reply:
code\datums\EPv2.dm:99:error: exonet.address: undefined var
code\datums\EPv2.dm:108:error: exonet.address: undefined var
code\datums\EPv2.dm:109:error: returnexonet.holder: undefined var

No, this should be type loop.